### PR TITLE
Made ref pattern bindings correctly pick Deref or DerefMut

### LINKED ISSUE #15609: Autoderef chooses deref() when deref_mut() is needed for ref mut pattern
```rust
use std::cell::RefCell;

struct S {
    node: E,
}

enum E {
    Foo(uint),
    Bar,
}

fn main() {
    let x = RefCell::new(S { node: Foo(0) });

    let mut b = x.borrow_mut();
    match b.node {
        Foo(ref mut n) => *n += 1,
        _ => (),
    }
}
```

```
$ rustc -v
rustc 0.11.0 (49bc17bfdd7143909aede81652d7d624cecd8a70 2014-07-07 17:16:34 +0000)
$ rustc foo.rs
foo.rs:17:13: 17:22 error: cannot borrow immutable anonymous field as mutable
foo.rs:17         Foo(ref mut n) => *n += 1,
                      ^~~~~~~~~
```

Changing to `match b.deref_mut().node` fixes the error.

### PR DESCRIPTION
**Pull request by Kimundi**
Closes #15609

### CODE REVIEW

**Comment by Kimundi** [comment]
r? @nikomatsakis 

cc @eddyb

**Comment by nikomatsakis** [comment]
r+ modulo the two nits

**Comment by Kimundi** [comment]
@nikomatsakis re the tests: I only realized after I wrote them that I did an unnecessary amount of duplication. If I'd used a `Cell` field instead of just a `panic`, it could probably all be one test that ensures it got never touched.

I'll change that, plus the` LValuePref` stuff.

**Comment by eddyb** [comment]
@nikomatsakis There is a subtle reason for not having `struct LvaluePreference(pub ast::Mutability)` to begin with, though it's not much of one: in the non-mutable case, immutable is actually *not* preferred, there is no preference.
May I suggest `struct LvalueUsage(pub ast::Mutability)`?

**Comment by Kimundi** [comment]
Fwiw, I just added the `from_mutbl` method, since there are only 3 places where the direct conversion is used, but ca 20 places where one of the two constants are hardcoded.

**Comment by Kimundi** [comment]
@nikomatsakis: Okay, I addressed both nits.

- There is now `LvaluePreference::from_mutbl()`, since that operation was less common that having one of the two as a constant.
- There are only two test files now, one for the issue scenario, and one testing all possible legal combinations of ref mutability and deref operator.

**Comment by nikomatsakis** [comment]
@eddyb true, the name is somewhat misleading.

File: src/librustc/middle/pat_util.rs [hunk 1]
```diff
@@ -135,12 +135,19 @@ pub fn pat_contains_bindings(dm: &DefMap, pat: &ast::Pat) -> bool {
     contains_bindings
 }
 
-/// Checks if the pattern contains any `ref` or `ref mut` bindings.
-pub fn pat_contains_ref_binding(dm: &DefMap, pat: &ast::Pat) -> bool {
-    let mut result = false;
+/// Checks if the pattern contains any `ref` or `ref mut` bindings,
+/// and if yes wether its containing mutable ones or just immutables ones.
+pub fn pat_contains_ref_binding(dm: &DefMap, pat: &ast::Pat) -> Option<ast::Mutability> {
```
Comment (jakub-) [inline src/librustc/middle/pat_util.rs]: Naming issue: function name still implies it's a predicate function but now it returns an Option. Same below.

**Pull request merged**